x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java

```diff
@@ -104,31 +91,21 @@ public boolean isSafeAction() {
     @Override
     public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
         StepKey checkNotWriteIndex = new StepKey(phase, NAME, CheckNotDataStreamWriteIndexStep.NAME);
+        StepKey waitForNoFollowerStepKey = new StepKey(phase, NAME, WaitForNoFollowersStep.NAME);
         StepKey readOnlyKey = new StepKey(phase, NAME, ReadOnlyStep.NAME);
         StepKey generateRollupIndexNameKey = new StepKey(phase, NAME, GENERATE_ROLLUP_STEP_NAME);
         StepKey rollupKey = new StepKey(phase, NAME, NAME);
         CheckNotDataStreamWriteIndexStep checkNotWriteIndexStep = new CheckNotDataStreamWriteIndexStep(checkNotWriteIndex, readOnlyKey);
+        WaitForNoFollowersStep waitForNoFollowersStep = new WaitForNoFollowersStep(waitForNoFollowerStepKey, readOnlyKey, client);
         ReadOnlyStep readOnlyStep = new ReadOnlyStep(readOnlyKey, generateRollupIndexNameKey, client);
         GenerateUniqueIndexNameStep generateRollupIndexNameStep = new GenerateUniqueIndexNameStep(
             generateRollupIndexNameKey,
             rollupKey,
             ROLLUP_INDEX_PREFIX,
             (rollupIndexName, lifecycleStateBuilder) -> lifecycleStateBuilder.setRollupIndexName(rollupIndexName)
         );
-        if (rollupPolicy == null) {
-            Step rollupStep = new RollupStep(rollupKey, nextStepKey, client, config);
-            return List.of(checkNotWriteIndexStep, readOnlyStep, generateRollupIndexNameStep, rollupStep);
-        } else {
-            StepKey updateRollupIndexPolicyStepKey = new StepKey(phase, NAME, UpdateRollupIndexPolicyStep.NAME);
-            Step rollupStep = new RollupStep(rollupKey, updateRollupIndexPolicyStepKey, client, config);
-            Step updateRollupIndexPolicyStep = new UpdateRollupIndexPolicyStep(
-                updateRollupIndexPolicyStepKey,
-                nextStepKey,
-                client,
-                rollupPolicy
-            );
-            return List.of(checkNotWriteIndexStep, readOnlyStep, generateRollupIndexNameStep, rollupStep, updateRollupIndexPolicyStep);
-        }
+        Step rollupStep = new RollupStep(rollupKey, nextStepKey, client, config);
```

Comment:
I wonder if there are more steps, that should be added before or after the rollup step.

Reply:
How will this work with TSDB data streams writing to indices *after* they have been rolled over? I can imagine that we'd want to wait until the index is out of the "valid" time range before doing the downsample, correct? Otherwise we risk documents getting rejected because the index is marked as read-only.

Reply:
Also, it doesn't appear that ILM re-adds the new index to the abstraction that it is currently in (the alias or data stream). Does the rollup action automatically do that, or do we need to add a step to do that somewhere?

Reply:
> How will this work with TSDB data streams writing to indices after they have been rolled over? I can imagine that we'd want to wait until the index is out of the "valid" time range before doing the downsample, correct? Otherwise we risk documents getting rejected because the index is marked as read-only.

We leave this up to the user to configure the ILM policy in a way that there is enough time for late arrivals. After the index has been downsampled, it becomes read-only and no delayed entries are allowed. 


> Also, it doesn't appear that ILM re-adds the new index to the abstraction that it is currently in (the alias or data stream). Does the rollup action automatically do that, or do we need to add a step to do that somewhere?

If the source index is a backing index of a data stream, it is replaced by the rollup index. This happens at the [TransportRollupAction](https://github.com/elastic/elasticsearch/blob/f64c75dca608760c0470896316c3aa81ade93845/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java#L567-L571). We chose this approach instead of doing so at the ILM action, because we wanted all the rollup process to be available as an API endpoint that can be invoked by the user.

As for an alias, we assume that the index is read-only before being downsampled. So, there should not be any write alias pointing to this index. Do you think we should cover the case of a read alias?